### ISSUE TITLE
SIP presence / hints /subscriptions

### ISSUE THREAD

**Issue by vieridipaola** [InnovateAsterisk/Browser-Phone#333]
Hi,

My Browser-Phone users add contacts by ticking the "Subscribe to Device State Notificagtions" checkbox. These contacts/buddies have 4-digit extensions.

The contacts are other PJSIP endpoints that register to the same Asterisk server. They can register with UDP clients as PJSIP/XXXX or as webrtc clients as PJSIP/webXXXX. Usually only one of the two at any given time -- not both at the same time.

So, in Asterisk pjsip_wizard.conf I configured:

```
has_hint = yes
hint_context = default
```

and in extensions.conf I added this to the default context:

`exten => _XXXX,hint,PJSIP/${EXTEN}&PJSIP/web${EXTEN}`

The first time the contact/buddy is created its presence is not reflected neither in the left panel buddy list nor in the buddy full panel on the right. If I reload the Browser-Phone page then the buddy list on the left still does not reflect its status, but the full panel on the right which shows up when clicking on the buddy actually does.
I have to wait for a while - maybe until an event occurs (like when the contacts receives a call) - before the presence status is available on both the left-panel buddy list and elsewhere.

Why is that?

Do I need to review my Asterisk configuration, or is there something I need to do in Browser-Phone?

**Comment by InnovateAsterisk** [comment]
When the SUSBSCRIBE message gets sent, what is the NOTIFY response? (You can see this in the sip trace of the browser phone or the Asterisk server).

**Comment by vieridipaola** [comment]
In my custom  web_hook_on_register function I enable client-side SIP tracing like this:

`ua._transport.configuration.traceSip = true;`

I then see the following in the browser console (I'm trimming a few things out from the output and trying to show only one complete SUBSCRIBE/NOTIFY trace between 4053 and 4062, 4053 being the Browser-Phone extension wanting to know the state of 4062):

```
phone.js?1665405852:3728 Starting Subscribe of all (117) Extension Buddies...
phone.js?1665405852:3767 SUBSCRIBE: 4062@sip.mydomain.org
sip-0.20.0.min.js:2 sip.Subscription | Subscription undefined transitioned to NotifyWait
sip-0.20.0.min.js:2 sip.Transport | Sending WebSocket message:

SUBSCRIBE sip:4062@sip.mydomain.org SIP/2.0
Via: SIP/2.0/WSS 192.0.2.227;branch=z9hG4bK1693706
To: <sip:4062@sip.mydomain.org>
From: "Test" <sip:web4053@sip.mydomain.org>;tag=82gcrna8ta
CSeq: 1 SUBSCRIBE
Call-ID: 7mh8vc9sj008a1n7qsh8
Max-Forwards: 70
Accept: application/pidf+xml
Allow: ACK,BYE,CANCEL,INFO,INVITE,MESSAGE,NOTIFY,OPTIONS,PRACK,REFER,REGISTER,SUBSCRIBE
Event: presence
Expires: 300
Contact: <sip:51madj5s@192.0.2.227;transport=wss>
Supported: outbound
User-Agent: HMBrowserPhone
Content-Length: 0

====================

sip-0.20.0.min.js:2 sip.Transport | Received WebSocket text message:

SIP/2.0 401 Unauthorized
Via: SIP/2.0/WSS 192.0.2.227;rport=36856;received=127.0.0.1;branch=z9hG4bK7317788
Call-ID: 7mh8vsmkfqot2i2sg7gd
From: "Test" <sip:web4053@sip.mydomain.org>;tag=80e23e9h5p
To: <sip:web4053@sip.mydomain.org>;tag=z9hG4bK7317788
CSeq: 1 SUBSCRIBE
WWW-Authenticate: Digest realm="asterisk",nonce="1665405855/d9d8b4a11ddeeea4cd85ca075b76647a",opaque="197309e33c4d0def",algorithm=md5,qop="auth"
Server: Asterisk PBX 18.10.0
Content-Length:  0



sip-0.20.0.min.js:2 sip.Transport | Sending WebSocket message:

SUBSCRIBE sip:web4053@sip.mydomain.org SIP/2.0
Via: SIP/2.0/WSS 192.0.2.227;branch=z9hG4bK2131931
To: <sip:web4053@sip.mydomain.org>
From: "Test" <sip:web4053@sip.mydomain.org>;tag=80e23e9h5p
CSeq: 2 SUBSCRIBE
Call-ID: 7mh8vsmkfqot2i2sg7gd
Max-Forwards: 70
Authorization: Digest algorithm=MD5, username="web4053", realm="asterisk", nonce="1665405855/d9d8b4a11ddeeea4cd85ca075b76647a", uri="sip:web4053@sip.mydomain.org", response="a5654b3a784e3b99a72fae771d7ab136", opaque="197309e33c4d0def", qop=auth, cnonce="u8t35tu0u6l1", nc=00000001
Allow: ACK,BYE,CANCEL,INFO,INVITE,MESSAGE,NOTIFY,OPTIONS,PRACK,REFER,REGISTER,SUBSCRIBE
Event: message-summary
Expires: 300
Contact: <sip:51madj5s@192.0.2.227;transport=wss>
Supported: outbound
User-Agent: HMBrowserPhone
Content-Length: 0



sip-0.20.0.min.js:2 sip.Transport | Received WebSocket text message:

SIP/2.0 401 Unauthorized
Via: SIP/2.0/WSS 192.0.2.227;rport=36856;received=127.0.0.1;branch=z9hG4bK1693706
Call-ID: 7mh8vc9sj008a1n7qsh8
From: "Test" <sip:web4053@sip.mydomain.org>;tag=82gcrna8ta
To: <sip:4062@sip.mydomain.org>;tag=z9hG4bK1693706
CSeq: 1 SUBSCRIBE
WWW-Authenticate: Digest realm="asterisk",nonce="1665405855/d9d8b4a11ddeeea4cd85ca075b76647a",opaque="1840149f2a6a64d5",algorithm=md5,qop="auth"
Server: Asterisk PBX 18.10.0
Content-Length:  0



sip-0.20.0.min.js:2 sip.Transport | Sending WebSocket message:

SUBSCRIBE sip:4062@sip.mydomain.org SIP/2.0
Via: SIP/2.0/WSS 192.0.2.227;branch=z9hG4bK6505207
To: <sip:4062@sip.mydomain.org>
From: "Test" <sip:web4053@sip.mydomain.org>;tag=82gcrna8ta
CSeq: 2 SUBSCRIBE
Call-ID: 7mh8vc9sj008a1n7qsh8
Max-Forwards: 70
Authorization: Digest algorithm=MD5, username="web4053", realm="asterisk", nonce="1665405855/d9d8b4a11ddeeea4cd85ca075b76647a", uri="sip:4062@sip.mydomain.org", response="5f75e79a17c0a3ccff97faa17e55ac11", opaque="1840149f2a6a64d5", qop=auth, cnonce="0s5nu1n0of24", nc=00000001
Accept: application/pidf+xml
Allow: ACK,BYE,CANCEL,INFO,INVITE,MESSAGE,NOTIFY,OPTIONS,PRACK,REFER,REGISTER,SUBSCRIBE
Event: presence
Expires: 300
Contact: <sip:51madj5s@192.0.2.227;transport=wss>
Supported: outbound
User-Agent: HMBrowserPhone
Content-Length: 0

===================

sip-0.20.0.min.js:2 sip.Transport | Received WebSocket text message:

NOTIFY sip:51madj5s@127.0.0.1:36856;transport=ws SIP/2.0
Via: SIP/2.0/WSS 127.0.0.1:8090;rport;branch=z9hG4bKPj65edffe0-be5f-4abb-be6c-406d86136f53;alias
From: <sip:4062@sip.mydomain.org>;tag=2ec23022-0b02-40bc-8f3e-a07d8d6eccfc
To: "Test" <sip:web4053@sip.mydomain.org>;tag=82gcrna8ta
Contact: <sip:127.0.0.1:8090;transport=ws>
Call-ID: 7mh8vc9sj008a1n7qsh8
CSeq: 4145 NOTIFY
Event: presence
Subscription-State: active;expires=299
Allow-Events: message-summary, presence, dialog, refer
Max-Forwards: 70
User-Agent: Asterisk PBX 18.10.0
Content-Type: application/pidf+xml
Content-Length:   413

<?xml version="1.0" encoding="UTF-8"?>
<presence entity="sip:4062@127.0.0.1:8090;transport=ws" xmlns="urn:ietf:params:xml:ns:pidf" xmlns:dm="urn:ietf:params:xml:ns:pidf:data-model" xmlns:rpid="urn:ietf:params:xml:ns:pidf:rpid">
 <note>Ready</note>
 <tuple id="4062">
  <status>
   <basic>open</basic>
  </status>
  <contact priority="1">sip:web4053@sip.mydomain.org</contact>
 </tuple>
 <dm:person />
</presence>
```


Any clues?

**Comment by vieridipaola** [comment]
I also noticed that if I select a buddy and press the edit icon then SAVE even without really editing anything I get the desired effect of seeing the endpoint's status.

**Comment by InnovateAsterisk** [comment]
The save function unsubscribes the buddy, and resubscribes them again regardless of changes. 

Question is why does this not happen the first time? I have tried this many times on my side, and it works fine, but I'm wondering is there is something with the pjsip_wizard - i have not used this option before... however your SIP trace does seem to be correct. Was this captured at a point when it was NOT working as expected?

**Comment by vieridipaola** [comment]
Your comment on the Save function put me on the right track.

It just so happens that the last thing Browser-Phone does is run a custom js code where I update the buddy list if there's a new one provisioned by my server. So, on successful download it populates the buddy list. I never noticed that before, but it "resets" the UI state of that extension until an event occurs that refreshes it. So for a short time lapse the presence state is unknown.

All I had to do to fix this was to add the following in my custom code right after loading/populating the buddy list:

```
PopulateBuddyList();
for(var b = 0; b < Buddies.length; b++) {
	if(Buddies[b].EnableSubscribe == true) {
		if (Buddies[b].ExtNo) console.log("Subscribe ", Buddies[b]
		SubscribeBuddy(Buddies[b]);
	}
}
```

My bad. Sorry for the confusion.
Closing issue shamefully...
